node/sc/subbridge_test.go

```diff
@@ -0,0 +1,326 @@
+	"fmt"
+	"github.com/golang/mock/gomock"
+	"github.com/klaytn/klaytn/accounts"
+	"github.com/klaytn/klaytn/crypto"
+	"github.com/klaytn/klaytn/event"
+	"github.com/klaytn/klaytn/networks/p2p"
+	"github.com/klaytn/klaytn/networks/p2p/discover"
+	"github.com/klaytn/klaytn/node"
+	"github.com/stretchr/testify/assert"
+	"math/big"
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+// testNewSubBridge returns a test SubBridge.
+func testNewSubBridge(t *testing.T) *SubBridge {
+	sCtx := node.NewServiceContext(&node.DefaultConfig, map[reflect.Type]node.Service{}, &event.TypeMux{}, &accounts.Manager{})
+	sBridge, err := NewSubBridge(sCtx, &SCConfig{NetworkId: testNetVersion, DataDir: testDataDir})
+	if err != nil {
+		t.Fatal(err)
+	}
+	assert.NotNil(t, sBridge)
+
+	return sBridge
+}
+
+// removeData removes blockchain data generated during the SubBridge test.
+func (sb *SubBridge) removeData(t *testing.T) {
+	if err := os.RemoveAll(sb.config.DataDir); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestSubBridge_basic tests some getters and basic operation of SubBridge.
+func TestSubBridge_basic(t *testing.T) {
+	// Create a test SubBridge
+	sBridge := testNewSubBridge(t)
+	defer sBridge.removeData(t)
+
+	// APIs returns default rpc APIs of MainBridge
+	apis := sBridge.APIs()
+	assert.Equal(t, 2, len(apis))
+	assert.Equal(t, "subbridge", apis[0].Namespace)
+	assert.Equal(t, "subbridge", apis[1].Namespace)
+
+	// Test getters for elements of SubBridge
+	assert.Equal(t, true, sBridge.IsListening()) // Always returns `true`
+	assert.Equal(t, testProtocolVersion, sBridge.ProtocolVersion())
+	assert.Equal(t, testNetVersion, sBridge.NetVersion())
+
+	// New components of MainBridge which will update old components
+	bc := testBlockChain(t)
+	txPool := testTxPool(bc)
+
+	var comp []interface{}
+	comp = append(comp, bc)
+	comp = append(comp, txPool)
+
+	// Check initial status of components
+	assert.Nil(t, sBridge.blockchain)
+	assert.Nil(t, sBridge.txPool)
+
+	// Update and check MainBridge components
+	sBridge.SetComponents(comp)
+	assert.Equal(t, bc, sBridge.blockchain)
+	assert.Equal(t, txPool, sBridge.txPool)
+
+	// Start MainBridge and stop later
+	if err := sBridge.Start(p2p.SingleChannelServer{}); err != nil {
+		t.Fatal(err)
+	}
+	defer sBridge.Stop()
+
+	//TODO more test
+}
+
+// TestSubBridge_removePeer tests correct removal of a peer from `SubBridge.peers`.
+func TestSubBridge_removePeer(t *testing.T) {
+	// Create a test SubBridge (it may have 0 peers)
+	sBridge := testNewSubBridge(t)
+	defer sBridge.chainDB.Close()
+	defer sBridge.removeData(t)
+
+	// Set components of SubBridge
+	bc := testBlockChain(t)
+	txPool := testTxPool(bc)
+
+	var comp []interface{}
+	comp = append(comp, bc)
+	comp = append(comp, txPool)
+
+	sBridge.SetComponents(comp)
+
+	// Prepare information of bridgePeer to be added and removed
+	nodeID := "0x1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d"
```

Comment:
Didn't you use this variable in another test? If this is reused, how about declaring a variable?

Reply:
Thank you for the suggestion. The string is currently being used in only two test function:  `TestMainBridge_removePeer`, `TestSubBridge_removePeer`. I think it will not be used in other tests anymore. I will declare it as global when it is used in more tests.